src/transaction/log_append.cpp

```diff
@@ -0,0 +1,1868 @@
+	}
+      else
+	{
+	  if (rcvindex == RVES_NOTIFY_VACUUM)
+	    {
+	      VFID_SET_NULL (&vacuum_info_p->vfid);
+	    }
+	  else
+	    {
+	      /* We require VFID for vacuum */
+	      assert_release (false);
+	      error_code = ER_FAILED;
+	      goto error;
+	    }
+	}
+
+      /* Initialize previous MVCC op log lsa - will be completed later */
+      LSA_SET_NULL (&vacuum_info_p->prev_mvcc_op_log_lsa);
+    }
+
+  if (is_undo_zip)
+    {
+      assert (has_undo && (data_header_ulength_p != NULL));
+
+      *data_header_ulength_p = MAKE_ZIP_LEN (zip_undo->data_length);
+      error_code = prior_lsa_copy_undo_data_to_node (node, zip_undo->data_length, (char *) zip_undo->log_data);
+    }
+  else if (has_undo)
+    {
+      assert (data_header_ulength_p != NULL);
+
+      *data_header_ulength_p = ulength;
+      error_code = prior_lsa_copy_undo_crumbs_to_node (node, num_ucrumbs, ucrumbs);
+    }
+
+  if (is_redo_zip)
+    {
+      assert (has_redo && (data_header_rlength_p != NULL));
+
+      *data_header_rlength_p = MAKE_ZIP_LEN (zip_redo->data_length);
+      error_code = prior_lsa_copy_redo_data_to_node (node, zip_redo->data_length, (char *) zip_redo->log_data);
+    }
+  else if (has_redo)
+    {
+      *data_header_rlength_p = rlength;
+      error_code = prior_lsa_copy_redo_crumbs_to_node (node, num_rcrumbs, rcrumbs);
+    }
+
+  if (error_code != NO_ERROR)
+    {
+      goto error;
+    }
+
+  return error_code;
+
+error:
+  if (node->data_header != NULL)
+    {
+      free_and_init (node->data_header);
+    }
+  if (node->udata != NULL)
+    {
+      free_and_init (node->udata);
+    }
+  if (node->rdata != NULL)
+    {
+      free_and_init (node->rdata);
+    }
+
+  return error_code;
+}
+
+/*
+ * prior_lsa_gen_postpone_record -
+ *
+ * return: error code or NO_ERROR
+ *
+ *   node(in/out):
+ *   rcvindex(in):
+ *   addr(in):
+ *   length(in):
+ *   data(in):
+ */
+static int
+prior_lsa_gen_postpone_record (THREAD_ENTRY *thread_p, LOG_PRIOR_NODE *node, LOG_RCVINDEX rcvindex,
+			       LOG_DATA_ADDR *addr, int length, const char *data)
+{
+  LOG_REC_REDO *redo;
+  VPID *vpid;
+  int error_code = NO_ERROR;
+
+  node->data_header_length = sizeof (LOG_REC_REDO);
+  node->data_header = (char *) malloc (node->data_header_length);
+  if (node->data_header == NULL)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, (size_t) node->data_header_length);
+      return ER_OUT_OF_VIRTUAL_MEMORY;
+    }
+  redo = (LOG_REC_REDO *) node->data_header;
+
+  redo->data.rcvindex = rcvindex;
+  if (addr->pgptr != NULL)
+    {
+      vpid = pgbuf_get_vpid_ptr (addr->pgptr);
+      redo->data.pageid = vpid->pageid;
+      redo->data.volid = vpid->volid;
+    }
+  else
+    {
+      redo->data.pageid = NULL_PAGEID;
+      redo->data.volid = NULL_VOLID;
+    }
+  redo->data.offset = addr->offset;
+
+  redo->length = length;
+  error_code = prior_lsa_copy_redo_data_to_node (node, redo->length, data);
```

Comment:
if error, free data_header

Reply:
node->data_header is freed by prior_lsa_alloc_and_copy_data when error occurs.